docs/utilities/data_masking.md

```diff
@@ -0,0 +1,578 @@
+        ```
+
+    === "Result"
+
+        ```json hl_lines="16"
+        --8<-- "examples/data_masking/src/choosing_payload_list_slice_output.json"
+        ```
+
+=== "Complex expressions"
+
+    You want to obfuscate data by finding for a field with conditional expression.
+
+    === "Data"
+
+        > Expression: `data_masker.mask(data, fields=["$.address[?(@.postcode > 81846)]"])`
+
+        > `$`: Represents the root of the JSON structure.
+
+        > `.address`: Selects the "address" property within the JSON structure.
+
+        > `(@.postcode > 81846)`: Specifies the condition that elements should meet. It selects elements where the value of the `postcode` property is `greater than 81846`.
+
+        ```json hl_lines="8 12"
+        --8<-- "examples/data_masking/src/choosing_payload_complex_search.json"
+        ```
+
+    === "Result"
+
+        ```json hl_lines="8 12"
+        --8<-- "examples/data_masking/src/choosing_payload_complex_search_output.json"
+        ```
+For comprehensive guidance on using JSONPath syntax, please refer to the official documentation available at [jsonpath-ng](https://github.com/h2non/jsonpath-ng#jsonpath-syntax){target="_blank" rel="nofollow"}
+
+#### JSON
+
+We also support data in JSON string format as input. We automatically deserialize it, then handle each field operation as expected.
+
+Note that the return will be a deserialized JSON and your desired fields updated.
+
+=== "Data"
+
+    Expression: `data_masker.mask(data, fields=["card_number", "address.postcode"])`
+
+    ```json
+    --8<-- "examples/data_masking/src/choosing_payload_simple_json.json"
+    ```
+
+=== "Result"
+
+    ```json
+    --8<-- "examples/data_masking/src/choosing_payload_simple_json_output.json"
+    ```
+
+## Advanced
+
+### Data serialization
+
+???+ note "Current limitations"
+    1. Python classes, `Dataclasses`, and `Pydantic models` are not supported yet.
+
+Before we traverse the data structure, we perform two important operations on input data:
+
+1. If `JSON string`, **deserialize** using default or provided deserializer.
+2. If `dictionary`, **normalize** into `JSON` to prevent traversing unsupported data types.
+
+When decrypting, we revert the operation to restore the original data structure.
+
+For compatibility or performance, you can optionally pass your own JSON serializer and deserializer to replace `json.dumps` and `json.loads` respectively:
+
+=== "advanced_custom_serializer.py"
+
+    ```python hl_lines="16"
+    --8<-- "examples/data_masking/src/advanced_custom_serializer.py"
+    ```
+
+### Providers
+
+#### AWS Encryption SDK
+
+You can modify the following values when initializing the `AWSEncryptionSDKProvider` to best accommodate your security and performance thresholds.
+
+| Parameter                  | Required | Default               | Description                                                                                   |
```

Comment:
Are any of these parameters required? if not, there's no need to have the column